src/bitgraphs.jl

```diff
@@ -0,0 +1,117 @@
+    bitgraph::Vector{INT} # store the adjacency matrix as vector of BitStr
+    fadjlist::Vector{Vector{Int}} # store the adjacency list, informate of sparse graph
+    mask::INT # mask for the graph
+end
+
+function BitGraph(g::SimpleGraph)
+    N = nv(g)
+    INT = LongLongUInt{N ÷ 64 + 1}
+    bitgraph = [bmask(INT, g.fadjlist[i]) for i in 1:N]
+    mask = bmask(INT, 1:N)
+    return BitGraph(N, bitgraph, deepcopy(g.fadjlist), mask)
+end
+
+function Graphs.nv(bg::BitGraph{INT}) where{INT}
+    return count_ones(bg.mask)
+end
+
+Base.show(io::IO, bg::BitGraph{INT}) where {INT} = print(io, "BitGraph{$INT}, N: $(bg.N), nv: $(nv(bg)), mask: $(BitStr{bg.N}(bg.mask))")
+Base.copy(bg::BitGraph{INT}) where{INT} = BitGraph(bg.N, copy(bg.bitgraph), deepcopy(bg.fadjlist), bg.mask)
+
+function Graphs.neighbors(bg::BitGraph{INT}, v::Int) where{INT}
+    return bg.bitgraph[v] & bg.mask
+end
+
+function closed_neighbors(bg::BitGraph{INT}, vs::INT) where{INT}
+    nebis = zero(INT)
+    for i in 1:bg.N
+        iszero(readbit(vs, i)) && continue
+        nebis |= bg.bitgraph[i]
+    end
+    return (nebis & bg.mask) | vs
+end
+
+function open_neighbors(bg::BitGraph{INT}, vs::INT) where{INT}
+    nebis = zero(INT)
+    for i in 1:bg.N
+        iszero(readbit(vs, i)) && continue
+        nebis |= bg.bitgraph[i]
+    end
+    return (nebis & bg.mask) & ~vs
+end
+
+function Graphs.is_connected(bg::BitGraph{INT}, v::Int, u::Int) where{INT}
+    @assert (readbit(bg.mask, v) == 1) && (readbit(bg.mask, u) == 1)
+    return isone(readbit(bg.bitgraph[v], u))
+end
+
+function Graphs.add_edge!(bg::BitGraph{INT}, src::Int, dst::Int) where{INT}
+    bg.bitgraph[src] = bg.bitgraph[src] | bmask(INT, dst)
+    bg.bitgraph[dst] = bg.bitgraph[dst] | bmask(INT, src)
+    push!(bg.fadjlist[src], dst)
+    push!(bg.fadjlist[dst], src)
+    nothing
+end
+
+# consider induced subgraph with mask
+function Graphs.induced_subgraph(bg::BitGraph{INT}, vertices::INT) where{INT}
+    sub_bg = copy(bg)
+    sub_bg.mask = sub_bg.mask & vertices
+    return sub_bg
+end
+
+# we actually need neighobors of connected components of the induced subgraph
+
+# mask is a parameter, to aviod creating new bitgraph when considering induced subgraph
+function Graphs.connected_components(bg::BitGraph{INT}; mask::INT = bg.mask) where{INT}
+    comps = Vector{INT}()
+    labels = ~mask
+
+    for i in 1:bg.N
+        readbit(bg.mask, i) == 0 && continue
+        readbit(labels, i) != 0 && continue
+        comp = bmask(INT, i)
+        labels = labels | comp
+        comp, labels = _connected_components(bg, comp, labels, i)
+        push!(comps, comp & bg.mask)
+    end
+    return comps
+end
+
+function _connected_components(bg::BitGraph{INT}, comp::INT, labels::INT, i::Int) where{INT}
+    active_nebis = (neighbors(bg, i) & ~labels)
+    (active_nebis == 0) && return comp, labels
+    comp = comp | active_nebis
+    labels = labels | active_nebis
+    for j in bg.fadjlist[i]
```

Comment:
Yes, since `bit_connected_components` is the most frequently function, and storing the `fadjlist` allow the function to be $O(Nd)$ instead of $O(N^2)$, where $d$ is degree of the graph.
Here is an example
```Julia
julia> using TreeWidthSolver, Graphs, BenchmarkTools

julia> for n in 10:10:50
           g = random_regular_graph(n, 3)
           bg = MaskedBitGraph(g)
           tg = @belapsed connected_components($g)
           tbg = @belapsed TreeWidthSolver.bit_connected_components($bg)
           @show n, tg, tbg
       end
(n, tg, tbg) = (10, 3.609095238095238e-7, 8.058101135190918e-8)
(n, tg, tbg) = (20, 5.518095238095239e-7, 1.341282633371169e-7)
(n, tg, tbg) = (30, 7.423664122137404e-7, 1.7956204379562043e-7)
(n, tg, tbg) = (40, 9.198717948717949e-7, 2.3305714285714285e-7)
(n, tg, tbg) = (50, 1.1416e-6, 2.893579335793358e-7)
```